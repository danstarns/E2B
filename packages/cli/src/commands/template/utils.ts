import * as e2b from 'e2b'
import { asLocalRelative } from 'src/utils/format'

import { client } from 'src/api'

const requestTemplateBuild = e2b.withAccessToken(
  client.api.path('/templates').method('post').create()
)

const requestTemplateRebuild = e2b.withAccessToken(
  client.api.path('/templates/{templateID}').method('post').create()
)

export async function requestBuildTemplate(
  accessToken: string,
  args: e2b.paths['/templates']['post']['requestBody']['content']['application/json'],
  hasConfig: boolean,
  configPath: string,
  templateID?: string
): Promise<
  Omit<
    e2b.paths['/templates']['post']['responses']['202']['content']['application/json'],
    'logs'
  >
> {
  let res
  if (templateID) {
    res = await requestTemplateRebuild(accessToken, { templateID, ...args })
  } else {
    res = await requestTemplateBuild(accessToken, args)
  }

  if (!res.ok) {
    const error:
      | e2b.paths['/templates']['post']['responses']['401']['content']['application/json']
      | e2b.paths['/templates']['post']['responses']['500']['content']['application/json'] =
      res.data as any

    if (error.code === 401) {
      throw new Error(
        `Authentication error: ${res.statusText}, ${
          error.message ?? 'no message'
        }`
      )
    }

    if (error.code === 404) {
      throw new Error(
        `Sandbox template you want to build ${
          templateID ? `(${templateID})` : ''
        } not found: ${res.statusText}, ${error.message ?? 'no message'}\n${
          hasConfig
            ? `This could be caused by ${asLocalRelative(
                configPath
              )} belonging to a deleted template or a template that you don't own. If so you can delete the ${asLocalRelative(
                configPath
              )} and start building the template again.`
            : ''
        }`
      )
    }

    if (error.code === 500) {
      throw new Error(
        `Server error: ${res.statusText}, ${error.message ?? 'no message'}`
      )
    }

    throw new Error(
      `API request failed: ${res.statusText}, ${error.message ?? 'no message'}`
    )
  }

  return res.data as any
}

const triggerTemplateBuild = e2b.withAccessToken(
  client.api
    .path('/templates/{templateID}/builds/{buildID}')
    .method('post')
    .create()
)

export async function triggerBuild(
  accessToken: string,
  templateID: string,
  buildID: string
) {
  const res = await triggerTemplateBuild(accessToken, { templateID, buildID })

  if (!res.ok) {
    const error:
      | e2b.paths['/templates/{templateID}/builds/{buildID}']['post']['responses']['401']['content']['application/json']
      | e2b.paths['/templates/{templateID}/builds/{buildID}']['post']['responses']['500']['content']['application/json'] =
      res.data as any

    if (error.code === 401) {
      throw new Error(
        `Authentication error: ${res.statusText}, ${
          error.message ?? 'no message'
        }`
      )
    }

    if (error.code === 500) {
      throw new Error(
        `Server error: ${res.statusText}, ${error.message ?? 'no message'}`
      )
    }

    throw new Error(
      `API request failed: ${res.statusText}, ${error.message ?? 'no message'}`
    )
  }

  return
}
