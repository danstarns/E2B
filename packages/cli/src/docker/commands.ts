import * as e2b from 'e2b'
import * as child_process from 'child_process'
import Docker from 'dockerode'

const docker = new Docker()

export function dockerConnect(accessToken: string) {
  return new Promise<void>((resolve) => {
    docker.checkAuth(
      {
        registry: `docker.${e2b.SANDBOX_DOMAIN}`,
        authconfig: {
          username: '_e2b_access_token',
          password: accessToken,
        },
      },
      (err, res) => {
        if (err) {
          console.error(
            'Docker login failed. Please try to log in with `e2b auth login` and try again.'
          )
          console.error(err)
          process.exit(1)
        } else {
          if (res.Status === 'Login Succeeded') {
            console.log('Docker login successful.')
            resolve()
          } else {
            console.error(`Docker login failed. ${res.Status}`)
            process.exit(1)
          }
        }
      }
    )
  })
}

export function dockerBuild(
  dockerfileRelativePath: string,
  templateID: string,
  template: Omit<
    {
      templateID: string;
      buildID: string;
      cpuCount: e2b.components['schemas']['CPUCount'];
      memoryMB: e2b.components['schemas']['MemoryMB'];
      public: boolean;
      aliases?: string[];
    },
    'logs'
  >,
  dockerBuildArgs: { [key: string]: string },
  root: string
) {
  console.log('Building docker image...')
  const cmd = `docker build . -f ${dockerfileRelativePath} --pull --platform linux/amd64 -t docker.${
    e2b.SANDBOX_DOMAIN
  }/e2b/custom-envs/${templateID}:${template.buildID} ${Object.entries(
    dockerBuildArgs
  )
    .map(([key, value]) => `--build-arg="${key}=${value}"`)
    .join(' ')}`
  child_process.execSync(cmd, {
    stdio: 'inherit',
    cwd: root,
    env: {
      ...process.env,
      DOCKER_CLI_HINTS: 'false',
    },
  })
  console.log('Docker image built.\n')
}

export function pushDockerImage(
  templateID: string,
  template: Omit<
    {
      templateID: string;
      buildID: string;
      cpuCount: e2b.components['schemas']['CPUCount'];
      memoryMB: e2b.components['schemas']['MemoryMB'];
      public: boolean;
      aliases?: string[];
    },
    'logs'
  >,
  root: string
) {
  console.log('Pushing docker image...')
  child_process.execSync(
    `docker push docker.${e2b.SANDBOX_DOMAIN}/e2b/custom-envs/${templateID}:${template.buildID}`,
    {
      stdio: 'inherit',
      cwd: root,
    }
  )
  console.log('Docker image pushed.\n')
}
