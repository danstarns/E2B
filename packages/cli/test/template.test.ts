import { test, expect, describe, vi } from 'vitest'
import { templateCommand } from '../src/commands/template'
import { runCommandWithOutput } from './run-command'
import { buildCommand } from 'src/commands/template/build'
import commandExists from 'command-exists'
import stripAnsi from 'strip-ansi'

describe('template', () => {
  test('should snapshot the --help output', async () => {
    const output = await runCommandWithOutput(templateCommand, [
      'template',
      '--help',
    ])

    expect(stripAnsi(output as string)).toMatchSnapshot()
  })

  describe('template build', () => {
    test('should shapshot help on the build command', async () => {
      const output = await runCommandWithOutput(buildCommand, [
        'build',
        '--help',
      ])

      expect(stripAnsi(output as string)).toMatchSnapshot()
    })

    test('should print an error and exit if Docker is not installed', async () => {
      vi.spyOn(commandExists, 'sync').mockReturnValue(false)

      const consoleErrorSpy = vi
        .spyOn(console, 'error')
        .mockImplementation(() => {})

      const processExitSpy = vi
        .spyOn(process, 'exit')
        .mockImplementation(() => {})

      await buildCommand.parseAsync(['build'])

      expect(consoleErrorSpy).toHaveBeenCalledWith(
        'Docker is required to build and push the sandbox template. Please install Docker and try again.'
      )
      expect(processExitSpy).toHaveBeenCalledWith(1)

      consoleErrorSpy.mockRestore()
      processExitSpy.mockRestore()

      vi.spyOn(commandExists, 'sync').mockReturnValue(true)
    })

    test('should print an error and exit if the name is invalid', async () => {
      const invalidName = '$$%%**'

      const consoleErrorSpy = vi
        .spyOn(console, 'error')
        .mockImplementation(() => {})

      const processExitSpy = vi
        .spyOn(process, 'exit')
        .mockImplementation(() => {})

      await buildCommand.parseAsync(['build', '--name', invalidName], {
        from: 'user',
      })

      const errorCalls = consoleErrorSpy.mock.calls
      const call = errorCalls[0][0]
      expect(stripAnsi(call)).toMatchInlineSnapshot(
        '"Name $$%%** is not valid. Name can only contain lowercase letters, numbers, dashes and underscores."'
      )

      expect(processExitSpy).toHaveBeenCalledWith(1)

      consoleErrorSpy.mockRestore()
      processExitSpy.mockRestore()
    })
  })
})
