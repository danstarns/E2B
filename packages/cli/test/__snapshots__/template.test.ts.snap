// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > should snapshot the --help output 1`] = `
"Usage: template [options] [command]

Options:
  -h, --help      display help for command

Commands:
  template|tpl    manage sandbox templates
  help [command]  display help for command
"
`;

exports[`template > template build > should shapshot help on the build command 1`] = `
"Usage: build [options] [command]

Options:
  -h, --help                     display help for command

Commands:
  build|bd [options] [template]  build sandbox template defined by ./e2b.Dockerfile or
                                 ./Dockerfile in root directory. By default the root directory is the
                                 current working directory. This command also creates e2b.toml config.
  help [command]                 display help for command
"
`;

exports[`template > template build > should throw if access token not found 1`] = `
"
                       ╭────────────────────────────────────────────────────────────────────╮
                       │ You must be logged in to use this command. Run e2b auth login.     │
                       │                                                                    │
                       │ If you are seeing this message in CI/CD you may need to set the    │
                       │ E2B_ACCESS_TOKEN environment variable.                             │
                       │ Visit https://e2b.dev/docs/getting-started/api-key to get the      │
                       │ access token.                                                      │
                       ╰────────────────────────────────────────────────────────────────────╯
"
`;
